### PR TITLE
ISPN-8074 failing random in computeIfPresent

### PR DESCRIPTION
**Pull request by karesti**
https://issues.jboss.org/browse/ISPN-8074

Corrected the L1NOnTxInterceptor when an exception is raised from
the command.
This is done by properly waiting flush cache method call and propagating the exception to avoid race conditions.

### CODE REVIEW

File: core/src/main/java/org/infinispan/interceptors/distribution/L1NonTxInterceptor.java [hunk 1]
```diff
@@ -359,21 +359,31 @@ private Object handleDataWriteCommand(InvocationContext ctx, DataWriteCommand co
          return invokeNext(ctx, command);
       }
       CompletableFuture<?> l1InvalidationFuture = invalidateL1InCluster(ctx, command, assumeOriginKeptEntryInL1);
-      return invokeNextThenApply(ctx, command, (rCtx, rCommand, rv) -> {
+      return invokeNextAndHandle(ctx, command, (InvocationContext rCtx, VisitableCommand rCommand, Object rv, Throwable ex) -> {
```
Comment (rvansa) [inline core/src/main/java/org/infinispan/interceptors/distribution/L1NonTxInterceptor.java]: please apply the same fix in `handleWriteManyCommand`, too

File: core/src/main/java/org/infinispan/interceptors/distribution/L1NonTxInterceptor.java [hunk 2]
```diff
@@ -359,21 +359,31 @@ private Object handleDataWriteCommand(InvocationContext ctx, DataWriteCommand co
...
+               });
             } else {
-               return asyncValue(l1InvalidationFuture.thenApply(nil -> rv));
+               throw ex;
+            }
+         } else {
+            if (l1InvalidationFuture != null && !l1InvalidationFuture.isDone() && isSynchronous(dataWriteCommand)) {
+               if (shouldRemoveFromLocalL1(rCtx, dataWriteCommand)) {
+                  VisitableCommand removeFromL1Command = removeFromL1Command(rCtx, dataWriteCommand.getKey());
+                  makeStage(asyncInvokeNext(rCtx, removeFromL1Command, l1InvalidationFuture))
```
Comment (rvansa) [inline core/src/main/java/org/infinispan/interceptors/distribution/L1NonTxInterceptor.java]: Missing `return` statement. Yes, the original code is missing that, too - that means that we haven't been waiting for the invalidation future to complete in the past either...

File: core/src/main/java/org/infinispan/interceptors/distribution/L1NonTxInterceptor.java [hunk 3]
```diff
@@ -359,21 +359,31 @@ private Object handleDataWriteCommand(InvocationContext ctx, DataWriteCommand co
...
+         if (ex != null) {
+            if (l1InvalidationFuture != null && !l1InvalidationFuture.isDone() && isSynchronous(dataWriteCommand) ) {
+               return asyncValue(l1InvalidationFuture).thenApply(rCtx, rCommand, (rCtx1, rCommand1, rv1) -> {
+                  throw ex;
+               });
             } else {
-               return asyncValue(l1InvalidationFuture.thenApply(nil -> rv));
+               throw ex;
+            }
+         } else {
```
Comment (rvansa) [inline core/src/main/java/org/infinispan/interceptors/distribution/L1NonTxInterceptor.java]: Cosmetic: the previous branch returns or throws, no need to increase nesting level using `else { ... }`...

File: core/src/main/java/org/infinispan/interceptors/distribution/L1NonTxInterceptor.java (hunk 2 above)
Comment (karesti) [inline core/src/main/java/org/infinispan/interceptors/distribution/L1NonTxInterceptor.java]: yes, I realised about that with Dan, and I was surprised that I was not entering this if in my tests, I'm going to add the return

File: core/src/main/java/org/infinispan/interceptors/distribution/L1NonTxInterceptor.java (hunk 3 above)
Comment (karesti) [inline core/src/main/java/org/infinispan/interceptors/distribution/L1NonTxInterceptor.java]: I don't understand, you mean that I should remove the throw ex ?

Comment (karesti) [inline core/src/main/java/org/infinispan/interceptors/distribution/L1NonTxInterceptor.java]: ok I understand what you mean

**Comment by karesti** [comment]
comment adressed @rvansa

**Comment by rvansa** [comment]
LGTM, just let the CI finish the night shift.

**Comment by tristantarrant** [comment]
@rvansa CI finished successfully

**Comment by rvansa** [comment]
Integrated, thanks.

**Pull request merged**